frontend-react/src/hooks/filters/UseCursorManager.ts

```diff
@@ -1,110 +1,95 @@
-    const addNextCursor = (val: string) => {
-        addCursorNoDuplicates(cursors.size, val);
-        checkNextPrev();
-    };
+type CursorReducer<Cursors, CursorAction> = (
+    cursors: Cursors,
+    action: CursorAction
+) => Cursors;
 
-    /* Handles navigating the cursor map */
-    const goTo = (i: number) => {
-        if (i >= cursors.size) {
-            setCurrentIndex(currentIndex);
-        } else {
-            setCurrentIndex(i);
-            checkNextPrev(i);
-        }
-    };
+interface CursorManager {
+    cursors: Cursors;
+    hasPrev: boolean;
+    hasNext: boolean;
+    update: React.Dispatch<CursorAction>;
+}
 
-    /* Can reset the cursor map and current index  */
-    const resetCursors = (newFirst?: string) => {
-        setCursors(
-            new Map<number, string>([[0, newFirst || firstCursor || ""]])
-        );
-        setCurrentIndex(0);
-    };
+const cursorReducer = (state: Cursors, action: CursorAction): Cursors => {
+    const { type, payload } = action;
+    switch (type) {
+        case CursorActionType.PAGE_UP:
+            /* Pages up by setting new current with state.next,
+             * sets next to an empty string since ADD_NEXT handles
+             * that, and adds state.current to the history. */
+            return {
+                current: state.next,
+                next: "",
+                history: [...state.history, state.current],
+            } as Cursors;
+        case CursorActionType.PAGE_DOWN:
+            /* Pages down by popping the last history item as current,
+             * sets next to state.current, and spreads the remaining
+             * history */
+            return {
+                current: state.history.pop(),
+                next: state.current,
+                history: [...state.history],
+            } as Cursors;
+        case CursorActionType.RESET:
+            /* Resets all values */
+            return {
+                current: payload || "",
+                next: "",
+                history: [],
+            } as Cursors;
+        case CursorActionType.ADD_NEXT:
+            /* Adds next cursor value */
+            return {
+                ...state,
+                next: payload || "",
+            };
+        default:
+            return state;
+    }
+};
 
-    /* Returning values and controllers as objects to keep them clean */
-    const values: ICursorValues = useMemo(() => {
-        return {
-            cursor: cursor,
-            cursors: cursors,
-            currentIndex: currentIndex,
-            hasPrev: hasPrev,
-            hasNext: hasNext,
-        };
-    }, [cursor, cursors, currentIndex, hasPrev, hasNext]);
+const useCursorManager = (firstCursor?: string) => {
+    const [cursors, cursorDispatch] = useReducer<
+        CursorReducer<Cursors, CursorAction>
+    >(cursorReducer, {
+        current: firstCursor || "",
+        next: "",
+        history: [],
+    });
+    const hasPrev = useMemo(() => {
```

Comment:
I'm curious if you need to use a memo here? Did you consider `useCallback`?

Reply:
Memos and callbacks are so similar that it's hard to tell what the difference would be in this case, but I think it could go either way!

Reply:
I _think_ `useCallback` is lighter weight since it doesn't record/cache the result, but I'm not 100% certain. I was more just curious if it was an intentional decision on your part. Consider issue resolved.